core/services/bulletprooftxmanager/eth_confirmer.go

```diff
@@ -0,0 +1,543 @@
+	"github.com/smartcontractkit/chainlink/core/store"
+	"github.com/smartcontractkit/chainlink/core/store/models"
+	"github.com/smartcontractkit/chainlink/core/store/orm"
+
+	gethCommon "github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	gethTypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/jinzhu/gorm"
+	"github.com/pkg/errors"
+)
+
+var (
+	// ErrExternalWalletUsedNonce is the error string we save if we come to the conclusion that the transaction nonce was used by an external account
+	ErrExternalWalletUsedNonce = "external wallet used nonce"
+
+	ethConfirmerAdvisoryLockClassID  = int32(1)
+	ethConfirmerAdvisoryLockObjectID = int32(0)
+)
+
+// EthConfirmer is a broad service which performs four different tasks in sequence on every new longest chain
+// Step 1: Mark that all currently pending transaction attempts were broadcast before this block
+// Step 2: Check pending transactions for receipts
+// Step 3: See if any transactions have exceeded the gas bumping block threshold and, if so, bump them
+// Step 4: Check confirmed transactions to make sure they are still in the longest chain (reorg protection)
+type EthConfirmer interface {
+	store.HeadTrackable
+}
+
+type ethConfirmer struct {
+	store             *store.Store
+	gethClientWrapper store.GethClientWrapper
+	config            orm.ConfigReader
+}
+
+func NewEthConfirmer(store *store.Store, config orm.ConfigReader) *ethConfirmer {
+	return &ethConfirmer{
+		store:             store,
+		gethClientWrapper: store.GethClientWrapper,
+		config:            config,
+	}
+}
+
+// Do nothing on connect, simply wait for the next head
+func (ec *ethConfirmer) Connect(*models.Head) error {
+	return nil
+}
+
+func (ec *ethConfirmer) Disconnect() {
+	// pass
+}
+
+func (ec *ethConfirmer) OnNewLongestChain(head models.Head) {
+	if ec.config.EnableBulletproofTxManager() {
+		if err := ec.ProcessHead(head); err != nil {
+			logger.Error(err)
+		}
+	}
+}
+
+// ProcessHead takes all required transactions for the confirmer on a new head
+func (ec *ethConfirmer) ProcessHead(head models.Head) error {
+	return withAdvisoryLock(ec.store, ethConfirmerAdvisoryLockClassID, ethConfirmerAdvisoryLockObjectID, func() error {
+		return ec.processHead(head)
+	})
+}
+
+// NOTE: This SHOULD NOT be run concurrently or it could behave badly
+func (ec *ethConfirmer) processHead(head models.Head) error {
+	if err := ec.SetBroadcastBeforeBlockNum(head.Number); err != nil {
+		return err
+	}
+
+	if err := ec.CheckForReceipts(); err != nil {
+		return err
+	}
+
+	if err := ec.BumpGasWhereNecessary(head.Number); err != nil {
+		return err
+	}
+
+	return ec.EnsureConfirmedTransactionsInLongestChain(head)
+}
+
+func (ec *ethConfirmer) SetBroadcastBeforeBlockNum(blockNum int64) error {
+	return ec.store.GetRawDB().Exec(
+		`UPDATE eth_tx_attempts SET broadcast_before_block_num = ? WHERE broadcast_before_block_num IS NULL AND state = 'broadcast'`,
```

Comment:
Use prepared statement parameter here.

Reply:
See previous comment about prepared statements for the state.